File: Calculator/android-native/app/src/main/java/com/qaantum/calculatorhub/calculators/NewFinanceCalculators.kt
package com.qaantum.calculatorhub.calculators

import kotlin.math.ln
import kotlin.math.log10
import kotlin.math.log2
import kotlin.math.pow

// NEW FINANCE CALCULATORS

// NPV Calculator
data class NPVResult(val npv: Double, val isProfitable: Boolean)
class NPVCalculator {
    fun calculate(initialInvestment: Double, cashFlows: List<Double>, discountRate: Double): NPVResult {
        val r = discountRate / 100
        var npv = -initialInvestment
        cashFlows.forEachIndexed { i, cf ->
            npv += cf / (1 + r).pow(i + 1)
        }
        return NPVResult(npv, npv >= 0)
    }
}

// IRR Calculator (Newton-Raphson approximation)
data class IRRResult(val irr: Double, val isValid: Boolean)
class IRRCalculator {
    fun calculate(initialInvestment: Double, cashFlows: List<Double>): IRRResult {
        val allCashFlows = listOf(-initialInvestment) + cashFlows
        var rate = 0.1 // Initial guess
        val maxIterations = 100
        val tolerance = 0.0001
        
        repeat(maxIterations) {
            var npv = 0.0
            var derivative = 0.0
            allCashFlows.forEachIndexed { j, cf ->
                npv += cf / (1 + rate).pow(j)
                if (j > 0) derivative -= j * cf / (1 + rate).pow(j + 1)
            }
            if (kotlin.math.abs(derivative) < 1e-10) return IRRResult(rate * 100, false)
            val newRate = rate - npv / derivative
            if (kotlin.math.abs(newRate - rate) < tolerance) {
                return IRRResult(newRate * 100, true)
            }
            rate = newRate
        }
        return IRRResult(rate * 100, true)
    }
}

// Down Payment Calculator
data class DownPaymentResult(val downPayment: Double, val loanAmount: Double, val monthlyPayment: Double)
class DownPaymentCalculator {
    fun calculate(purchasePrice: Double, downPaymentPercent: Double, interestRate: Double, loanTermYears: Int): DownPaymentResult {
        val downPayment = purchasePrice * (downPaymentPercent / 100)
        val loanAmount = purchasePrice - downPayment
        val monthlyRate = interestRate / 100 / 12
        val numPayments = loanTermYears * 12
        
        val monthlyPayment = if (monthlyRate == 0.0) {
            loanAmount / numPayments
        } else {
            loanAmount * (monthlyRate * (1 + monthlyRate).pow(numPayments)) / ((1 + monthlyRate).pow(numPayments) - 1)
        }
        return DownPaymentResult(downPayment, loanAmount, monthlyPayment)
    }
}

// Paycheck Calculator
data class PaycheckResult(
    val grossPay: Double, val federalTax: Double, val stateTax: Double,
    val socialSecurity: Double, val medicare: Double, val healthInsurance: Double,
    val retirement: Double, val totalDeductions: Double, val netPay: Double
)
class PaycheckCalculator {
    fun calculate(
        annualSalary: Double, payPeriod: String, federalRate: Double, stateRate: Double,
        ssRate: Double = 6.2, medicareRate: Double = 1.45, healthInsurance: Double = 0.0, retirementPercent: Double = 0.0
    ): PaycheckResult {
        val periodsPerYear = when (payPeriod) {
            "Weekly" -> 52; "Biweekly" -> 26; "Semi-Monthly" -> 24; "Monthly" -> 12; else -> 1
        }
        val grossPay = annualSalary / periodsPerYear
        val federal = grossPay * federalRate / 100
        val state = grossPay * stateRate / 100
        val ss = grossPay * ssRate / 100
        val medicare = grossPay * medicareRate / 100
        val retirement = grossPay * retirementPercent / 100
        val total = federal + state + ss + medicare + healthInsurance + retirement
        return PaycheckResult(grossPay, federal, state, ss, medicare, healthInsurance, retirement, total, grossPay - total)
    }
}

// CD Calculator
data class CDResult(val totalValue: Double, val interestEarned: Double)
class CDCalculator {
    fun calculate(deposit: Double, apy: Double, termMonths: Int, compoundingFrequency: Int = 365): CDResult {
        val t = termMonths / 12.0
        val r = apy / 100
        val total = deposit * (1 + r / compoundingFrequency).pow(compoundingFrequency * t)
        return CDResult(total, total - deposit)
    }
}

// Tip Split Calculator
data class TipSplitResult(val tipAmount: Double, val totalWithTip: Double, val perPersonAmount: Double, val perPersonTip: Double)
class TipSplitCalculator {
    fun calculate(billAmount: Double, tipPercent: Double, numberOfPeople: Int): TipSplitResult {
        val tip = billAmount * tipPercent / 100
        val total = billAmount + tip
        return TipSplitResult(tip, total, total / numberOfPeople, tip / numberOfPeople)
    }
}

File: Calculator/ios-native/CalculatorHub/Calculators/NewFinanceCalculators.swift
import Foundation

// NEW FINANCE CALCULATORS

// NPV Calculator
struct NPVResult { let npv: Double; let isProfitable: Bool }
class NPVCalculator {
    func calculate(initialInvestment: Double, cashFlows: [Double], discountRate: Double) -> NPVResult {
        let r = discountRate / 100
        var npv = -initialInvestment
        for (i, cf) in cashFlows.enumerated() { npv += cf / pow(1 + r, Double(i + 1)) }
        return NPVResult(npv: npv, isProfitable: npv >= 0)
    }
}

// IRR Calculator
struct IRRResult { let irr: Double; let isValid: Bool }
class IRRCalculator {
    func calculate(initialInvestment: Double, cashFlows: [Double]) -> IRRResult {
        let allCashFlows = [-initialInvestment] + cashFlows
        var rate = 0.1
        for _ in 0..<100 {
            var npv = 0.0; var derivative = 0.0
            for (j, cf) in allCashFlows.enumerated() {
                npv += cf / pow(1 + rate, Double(j))
                if j > 0 { derivative -= Double(j) * cf / pow(1 + rate, Double(j + 1)) }
            }
            if abs(derivative) < 1e-10 { return IRRResult(irr: rate * 100, isValid: false) }
            let newRate = rate - npv / derivative
            if abs(newRate - rate) < 0.0001 { return IRRResult(irr: newRate * 100, isValid: true) }
            rate = newRate
        }
        return IRRResult(irr: rate * 100, isValid: true)
    }
}

// Down Payment Calculator
struct DownPaymentResult { let downPayment: Double; let loanAmount: Double; let monthlyPayment: Double }
class DownPaymentCalculator {
    func calculate(purchasePrice: Double, downPaymentPercent: Double, interestRate: Double, loanTermYears: Int) -> DownPaymentResult {
        let downPayment = purchasePrice * (downPaymentPercent / 100)
        let loanAmount = purchasePrice - downPayment
        let monthlyRate = interestRate / 100 / 12
        let numPayments = Double(loanTermYears * 12)
        let monthlyPayment = monthlyRate == 0 ? loanAmount / numPayments : loanAmount * (monthlyRate * pow(1 + monthlyRate, numPayments)) / (pow(1 + monthlyRate, numPayments) - 1)
        return DownPaymentResult(downPayment: downPayment, loanAmount: loanAmount, monthlyPayment: monthlyPayment)
    }
}

// Paycheck Calculator
struct PaycheckResult { let grossPay, federalTax, stateTax, socialSecurity, medicare, healthInsurance, retirement, totalDeductions, netPay: Double }
class PaycheckCalculator {
    func calculate(annualSalary: Double, payPeriod: String, federalRate: Double, stateRate: Double, ssRate: Double = 6.2, medicareRate: Double = 1.45, healthInsurance: Double = 0, retirementPercent: Double = 0) -> PaycheckResult {
        let periods: Double = { switch payPeriod { case "Weekly": return 52; case "Biweekly": return 26; case "Semi-Monthly": return 24; case "Monthly": return 12; default: return 1 } }()
        let grossPay = annualSalary / periods
        let federal = grossPay * federalRate / 100; let state = grossPay * stateRate / 100
        let ss = grossPay * ssRate / 100; let med = grossPay * medicareRate / 100; let ret = grossPay * retirementPercent / 100
        let total = federal + state + ss + med + healthInsurance + ret
        return PaycheckResult(grossPay: grossPay, federalTax: federal, stateTax: state, socialSecurity: ss, medicare: med, healthInsurance: healthInsurance, retirement: ret, totalDeductions: total, netPay: grossPay - total)
    }
}

// CD Calculator
struct CDResult { let totalValue: Double; let interestEarned: Double }
class CDCalculator {
    func calculate(deposit: Double, apy: Double, termMonths: Int, compoundingFrequency: Int = 365) -> CDResult {
        let t = Double(termMonths) / 12.0; let r = apy / 100
        let total = deposit * pow(1 + r / Double(compoundingFrequency), Double(compoundingFrequency) * t)
        return CDResult(totalValue: total, interestEarned: total - deposit)
    }
}

// Tip Split Calculator
struct TipSplitResult { let tipAmount, totalWithTip, perPersonAmount, perPersonTip: Double }
class TipSplitCalculator {
    func calculate(billAmount: Double, tipPercent: Double, numberOfPeople: Int) -> TipSplitResult {
        let tip = billAmount * tipPercent / 100; let total = billAmount + tip
        return TipSplitResult(tipAmount: tip, totalWithTip: total, perPersonAmount: total / Double(numberOfPeople), perPersonTip: tip / Double(numberOfPeople))
    }
}
